File: blue-hour-sim.py
import pygame
import sys
import time
pygame.init()

size = width, height = 1000, 1000

screen = pygame.display.set_mode(size)

RED = (255, 0, 0)
GREEN = (0, 255, 0)
BLUE = (0, 0, 255)
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)

TOP_CUBE = pygame.Rect(450, 250, 100, 100)
MIDDLE_CUBE = pygame.Rect(450, 360, 100, 100)
BOTTOM_CUBE = pygame.Rect(450, 470, 100, 100)

TOP_COLOR = pygame.Color(0, 0, 255)
MIDDLE_COLOR = pygame.Color(0, 0, 255)
BOTTOM_COLOR = pygame.Color(0, 0, 255)

SENSOR_ONE = pygame.Rect(340, 600, 20, 20)
SENSOR_TWO = pygame.Rect(440, 650, 20, 20)
SENSOR_THREE = pygame.Rect(540, 650, 20, 20)
SENSOR_FOUR = pygame.Rect(640, 600, 20, 20)

sensors = [SENSOR_ONE, SENSOR_TWO, SENSOR_THREE, SENSOR_FOUR]

upslope = range(255)
downslope = list(reversed(upslope))
color_trails = [upslope, [0] * 255, downslope]
R=0
G=1
B=2

frames = [0, -200, -400]
targets = [0, -200, -400]

top_target = -400
middle_target = -200
bottom_target = 0

top_frame = -400
middle_frame = -200
bottom_frame = 0

def setup():
	pygame.draw.rect(screen, WHITE, SENSOR_ONE)
	pygame.draw.rect(screen, WHITE, SENSOR_TWO)
	pygame.draw.rect(screen, WHITE, SENSOR_THREE)
	pygame.draw.rect(screen, WHITE, SENSOR_FOUR)
	pygame.draw.rect(screen, TOP_COLOR, TOP_CUBE)
	pygame.draw.rect(screen, MIDDLE_COLOR, MIDDLE_CUBE)
	pygame.draw.rect(screen, BOTTOM_COLOR, BOTTOM_CUBE)

def redrawTrail(current_step):
	pygame.draw.rect(screen, BLACK, (100, 740, 255, 200))
	pygame.draw.line(screen, WHITE, (100 + current_step, 775), (100 + current_step, 925))
	for step, val in enumerate(color_trails[0]):
		pygame.draw.circle(screen, RED, (100 + step, 800 - val/10), 1)
	for step, val in enumerate(color_trails[1]):
		pygame.draw.circle(screen, GREEN, (100 + step, 850 - val/10), 1)
	for step, val in enumerate(color_trails[2]):
		pygame.draw.circle(screen, BLUE, (100 + step, 900 - val/10), 1)
	for step in range(255):
		pygame.draw.circle(screen, pygame.Color(color_trails[R][step], color_trails[G][step], color_trails[B][step]), (100 + step, 750), 1)

def moveFrames(frames, targets):
	frames[0] = frames[0] + (targets[0]-frames[0])/100
	return frames

setup()

FPS = 10
clock = pygame.time.Clock()

while True:
	milliseconds = clock.tick(FPS)
	print(milliseconds)
	for event in pygame.event.get():
		if event.type == pygame.QUIT: sys.exit()
		if event.type == pygame.MOUSEBUTTONUP:
			pos = pygame.mouse.get_pos()
			print pos
			for sensor in sensors:
				if sensor.collidepoint(pos):
					print('triggered')
					targets[0] += 200
	pygame.draw.rect(screen, pygame.Color(color_trails[R][frames[0]], color_trails[G][frames[0]], color_trails[B][frames[0]]), BOTTOM_CUBE)
	frames = moveFrames(frames, targets)
	print frames
	pygame.display.flip()
	redrawTrail(frames[0])

File: blue-hour-sim.js
upslope = Array.from(new Array(255), (x,i) => i);
everything = Array.from(new Array(255), (x,i) => 255);
nothing = Array.from(new Array(255), (x,i) => 0);
downslope = Array.from(new Array(255), (x,i) => 255 - i);
color_trails = [nothing.concat(upslope), nothing.concat(nothing), everything.concat(downslope)];

const R=0;
const G=1;
const B=2;

frames = [0, 100, 200];
targets = [0, 100, 200];
homePositions = [0, 100, 200];

eventTotal = 0;
shyRobotTimeoutID = 0;

/*const TOP = 0;
const MIDDLE = 1;
const BOTTOM = 2;*/

const SENSOR_ONE_X = 340;
const SENSOR_ONE_Y = 500;
const SENSOR_TWO_X = 440;
const SENSOR_TWO_Y = 550;
const SENSOR_THREE_X = 540;
const SENSOR_THREE_Y = 550;
const SENSOR_FOUR_X = 640;
const SENSOR_FOUR_Y = 500;

function drawTower() {
    fill('blue');
    var TOP_CUBE = rect(450, 150, 100, 100);
    var MIDDLE_CUBE = rect(450, 260, 100, 100);
    var BOTTOM_CUBE = rect(450, 370, 100, 100);
}

function drawSensors() {
    fill('white');
    ellipse(SENSOR_ONE_X, SENSOR_ONE_Y, 20, 20);
    ellipse(SENSOR_TWO_X, SENSOR_TWO_Y, 20, 20);
    ellipse(SENSOR_THREE_X, SENSOR_THREE_Y, 20, 20);
    ellipse(SENSOR_FOUR_X, SENSOR_FOUR_Y, 20, 20);
}

function drawTrail() {
    fill('BLACK');
    rect(0, 600, 1000, 200);

    stroke('WHITE');
    line(50 + frames[0], 600, 50 + frames[0], 650);
    line(50 + frames[1], 600, 50 + frames[1], 650);
    line(50 + frames[2], 600, 50 + frames[2], 650);

    stroke('RED');
    color_trails[R].forEach(function (item, index, array) {
        point(50 + index, 650 - item/10);
    });

    stroke('GREEN');
    color_trails[G].forEach(function (item, index, array) {
        point(50 + index, 650 - item/10);
    });

    stroke('BLUE');
    color_trails[B].forEach(function (item, index, array) {
        point(50 + index, 650 - item/10);
    });

    for (var step in color_trails[R]) {
        stroke(color(color_trails[R][step], color_trails[G][step], color_trails[B][step]));
        point(50 + int(step), 600);
    }
}

function approachTargets(t, f) {
    f.forEach(function (item, index, array) {
        f[index] = item + (t[index] - item)/10; 
    });
    return f;
}

function decayTargets(t, h) {
    t.forEach(function (item, index, array) {
        if (item > h[index]) {
            t[index] = item - 2;
        };
    });
    return t;
}

function drawTargetApproach() {
    stroke('WHITE');
    for(var i=100; i>0; i--) {
        var t = [[i], [i], [i]];
        var f = [[0], [0], [0]];
        f = approachTargets(t, f);
        point(i, 100-f[0]);
    }
}

function mouseClicked() {
    clearTimeout(shyRobotTimeoutID);
    shyRobotTimeoutID = setTimeout(shyRobot, 5000);

    var motion = false;
    eventTotal = eventTotal + 1;

    if(dist(mouseX, mouseY, SENSOR_ONE_X, SENSOR_ONE_Y) < 10){
        console.log('sensor 1 triggered');
        motion = true;
    }
    if(dist(mouseX, mouseY, SENSOR_TWO_X, SENSOR_TWO_Y) < 10){
        console.log('sensor 2 triggered');
        motion = true;
    }
    if(dist(mouseX, mouseY, SENSOR_THREE_X, SENSOR_THREE_Y) < 10){
        console.log('sensor 3 triggered');
        motion = true;
    }
    if(dist(mouseX, mouseY, SENSOR_FOUR_X, SENSOR_FOUR_Y) < 10){
        console.log('sensor 4 triggered');
        motion = true;
    }
    if(motion) {
        targets.forEach(function (item, index, array) {
            targets[index] = item + 150;
        });
    }
}

function translateToMorse(digit) {
    switch(digit) {
        case 0: return '00000';
        case 1: return '10000';
        case 2: return '11000';
        case 3: return '11100';
        case 4: return '11110';
        case 5: return '11111';
        case 6: return '01111';
        case 7: return '00111';
        case 8: return '00011';
        case 9: return '00001';
    }
}

function constructMorseString() {
    var morse = '';
    morse = morse.concat(translateToMorse(eventTotal % 10));
    if(eventTotal > 10) {
        morse = 'L'.concat(morse);
        morse = translateToMorse(int(eventTotal/10)).concat(morse);
    }
    morse = morse.concat('W');
    return morse;
}

function drawShyRobotData() {
    fill('white');
    textSize(20);
    text(eventTotal, 10, 750);
    text(constructMorseString(), 10, 775);
}

function shyRobot() {
    var morse = constructMorseString();
    if ( typeof shyRobot.MorseIndex == 'undefined' ) {
        shyRobot.MorseIndex = 0;
    }
    if(morse[shyRobot.MorseIndex] == '1') {
        dit();
    } else {
        dah();
    }
    shyRobot.MorseIndex++;
    shyRobotTimeoutID = setTimeout(shyRobot, 1000);
}

function dit() {
    console.log('dit');
}

function dah() {
    console.log('dah');
}

function setup() {
    createCanvas(1000, 800);
    frameRate(10)
    noStroke(); 
    background(0);
    drawSensors();
    drawTower();
    shyRobotTimeoutID = setTimeout(shyRobot, 5000);
}

function draw() {
    drawTrail();
    targets = decayTargets(targets, homePositions);
    frames = approachTargets(targets, frames);
    drawTargetApproach();
    noStroke();
    fill(color(color_trails[R][int(frames[0])], color_trails[G][int(frames[0])], color_trails[B][int(frames[0])]));
    rect(450, 150, 100, 100);
    fill(color(color_trails[R][int(frames[1])], color_trails[G][int(frames[1])], color_trails[B][int(frames[1])]));
    rect(450, 260, 100, 100);
    fill(color(color_trails[R][int(frames[2])], color_trails[G][int(frames[2])], color_trails[B][int(frames[2])]));
    rect(450, 370, 100, 100);
    drawShyRobotData();
}